File: python/platforms/bluesky.py
from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any

import requests

from core.models import InboundEvent, OutboundReply
from .base import PlatformAdapter
from .catalog import PLATFORM_CATALOG


class BlueskyAdapter(PlatformAdapter):
    """AT Protocol adapter for text posts; media URLs use text fallback."""

    capabilities = PLATFORM_CATALOG["bluesky"]

    def __init__(self, service_url: str | None = None, access_jwt: str | None = None, repo: str | None = None, session: requests.Session | None = None):
        self.service_url = (service_url or os.getenv("BLUESKY_SERVICE_URL", "https://bsky.social")).rstrip("/")
        self.access_jwt = access_jwt or os.getenv("BLUESKY_ACCESS_JWT")
        self.repo = repo or os.getenv("BLUESKY_REPO")
        if not self.access_jwt or not self.repo:
            raise ValueError("BLUESKY_ACCESS_JWT and BLUESKY_REPO are required")
        self.session = session or requests.Session()

    def parse_event(self, payload: Any, headers: dict[str, str] | None = None) -> InboundEvent:
        record = payload.get("record", payload)
        author = payload.get("author", payload.get("did", ""))
        text = record.get("text")
        if not author or text is None:
            raise ValueError("Bluesky post has no supported content")
        return InboundEvent(
            platform="bluesky",
            user_id=str(author),
            content_type="text",
            text=text,
            reply_to_uri=payload.get("uri"),
            reply_to_cid=payload.get("cid"),
        )

    def send_reply(self, event: InboundEvent, reply: OutboundReply) -> None:
        headers = {"Authorization": f"Bearer {self.access_jwt}", "Content-Type": "application/json"}
        for message in reply.messages[: self.capabilities.max_reply_items or len(reply.messages)]:
            record = {
                "text": message.text or message.media_url or "",
                "createdAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            }
            if event.reply_to_uri and event.reply_to_cid:
                record["reply"] = {
                    "root": {"uri": event.reply_to_uri, "cid": event.reply_to_cid},
                    "parent": {"uri": event.reply_to_uri, "cid": event.reply_to_cid},
                }
            response = self.session.post(
                f"{self.service_url}/xrpc/com.atproto.repo.createRecord",
                headers=headers,
                json={"repo": self.repo, "collection": "app.bsky.feed.post", "record": record},
                timeout=30,
            )
            response.raise_for_status()

File: php/src/bluesky.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function bluesky_parse_event(array $payload): array
{
    $record = $payload['record'] ?? $payload; $author = $payload['author'] ?? $payload['did'] ?? ''; if ($author === '' || !array_key_exists('text', $record)) throw new InvalidArgumentException('Bluesky post has no supported content'); return ['platform' => 'bluesky', 'user_id' => (string)$author, 'content_type' => 'text', 'text' => $record['text'], 'reply_to_uri' => $payload['uri'] ?? null, 'reply_to_cid' => $payload['cid'] ?? null];
}
function bluesky_send_reply(array $event, array $reply, string $serviceUrl, string $jwt, string $repo): void
{
    if ($serviceUrl === '' || $jwt === '' || $repo === '') throw new InvalidArgumentException('Bluesky service URL, JWT and repo are required'); foreach (array_slice($reply['messages'] ?? [], 0, 5) as $message) { if (($message['type'] ?? 'text') !== 'text') throw new InvalidArgumentException('Bluesky Blob upload is required before media replies'); $record = ['$type' => 'app.bsky.feed.post', 'text' => $message['text'] ?? '', 'createdAt' => gmdate('Y-m-d\TH:i:s.v\Z')]; if (!empty($event['reply_to_uri']) && !empty($event['reply_to_cid'])) $record['reply'] = ['root' => ['uri' => $event['reply_to_uri'], 'cid' => $event['reply_to_cid']], 'parent' => ['uri' => $event['reply_to_uri'], 'cid' => $event['reply_to_cid']]]; $ch = adapter_curl_init(rtrim($serviceUrl, '/').'/xrpc/com.atproto.repo.createRecord'); curl_setopt_array($ch, [CURLOPT_POST => true, CURLOPT_RETURNTRANSFER => true, CURLOPT_HTTPHEADER => ['Authorization: Bearer '.$jwt, 'Content-Type: application/json'], CURLOPT_POSTFIELDS => json_encode(['repo' => $repo, 'collection' => 'app.bsky.feed.post', 'record' => $record], JSON_THROW_ON_ERROR)]); $result = curl_exec($ch); $status = curl_getinfo($ch, CURLINFO_RESPONSE_CODE); if ($result === false || $status < 200 || $status >= 300) throw new RuntimeException('Bluesky API request failed'); curl_close($ch); }
}
